File: utils/refToolsClient.js
#!/usr/bin/env node
/**
 * Ref-Tools MCP Client for ACIMguide Agent System (Node.js)
 * Provides token-efficient access to technical documentation for all agents
 * Compatible with both CommonJS and ES modules
 */

const axios = require('axios');
const { URL } = require('url');

/**
 * Configuration class for Ref-Tools MCP client
 */
class RefToolsConfig {
    constructor(options = {}) {
        this.baseURL = options.baseURL || process.env.REF_MCP_URL || 
                      `http://${process.env.REF_MCP_HOST || 'localhost'}:${process.env.REF_MCP_PORT || 4102}`;
        this.port = options.port || parseInt(process.env.REF_MCP_PORT) || 4102;
        this.transport = options.transport || process.env.REF_MCP_TRANSPORT || 'http';
        this.timeout = options.timeout || 20000;
        this.maxRetries = options.maxRetries || 3;
        this.fallbackUrls = options.fallbackUrls || [
            'https://ref-tools-mcp.acimguide.com',
            'http://localhost:8080'
        ];
    }
}

/**
 * High-level client for Ref-Tools MCP service
 * Provides token-efficient documentation access for agents
 */
class RefToolsClient {
    constructor(config = null) {
        this.config = config || new RefToolsConfig();
        
        // Create axios instance with default headers
        this.axios = axios.create({
            timeout: this.config.timeout,
            headers: {
                'Content-Type': 'application/json',
                'User-Agent': 'ACIMguide-Agent-System/1.0'
            }
        });
    }

    /**
     * Make HTTP request to MCP server with fallback support
     */
    async _makeRequest(endpoint, payload = null, method = 'POST') {
        const urlsToTry = [this.config.baseURL, ...this.config.fallbackUrls];
        
        for (let attempt = 0; attempt < urlsToTry.length; attempt++) {
            try {
                const baseURL = urlsToTry[attempt];
                const url = new URL(endpoint, baseURL).toString();
                
                console.debug(`Attempting ${method} ${url} (attempt ${attempt + 1})`);
                
                let response;
                if (method.toUpperCase() === 'POST') {
                    response = await this.axios.post(url, payload);
                } else {
                    response = await this.axios.get(url);
                }
                
                return response.data;
                
            } catch (error) {
                console.warn(`Request to ${urlsToTry[attempt]} failed:`, error.message);
                
                if (attempt === urlsToTry.length - 1) {
                    throw new RefToolsConnectionError(
                        `Failed to connect to Ref-Tools MCP service after trying ${urlsToTry.length} endpoints`
                    );
                }
            }
        }
    }

    /**
     * Check if the MCP service is healthy and responsive
     */
    async healthCheck() {
        try {
            return await this._makeRequest('/health', null, 'GET');
        } catch (error) {
            return {
                status: 'unhealthy',
                error: error.message,
                timestamp: Date.now()
            };
        }
    }

    /**
     * Get documentation for a specific library or API
     * 
     * @param {string} library - Name of the library/API (e.g., 'firebase', 'openai', 'react')
     * @param {string} query - Specific search query within the documentation
     * @param {number} maxTokens - Maximum tokens to return (helps with cost optimization)
     * @returns {Promise<Object>} Dictionary containing relevant documentation snippets
     */
    async getDocs(library, query = '', maxTokens = 2000) {
        const payload = {
            library,
            query,
            max_tokens: maxTokens,
            format: 'json'
        };
        
        return await this._makeRequest('/docs', payload);
    }

    /**
     * Search for specific API endpoints and usage patterns
     * 
     * @param {string} apiName - Name of the API (e.g., 'Firebase Admin', 'OpenAI')
     * @param {string} endpointPattern - Pattern to search for (e.g., '/users', 'chat/completions')
     * @param {string} method - HTTP method filter (GET, POST, etc.)
     * @returns {Promise<Object>} API documentation and examples
     */
    async searchAPI(apiName, endpointPattern = '', method = '') {
        const payload = {
            api: apiName,
            endpoint: endpointPattern,
            method: method.toUpperCase(),
            include_examples: true
        };
        
        return await this._makeRequest('/api', payload);
    }

    /**
     * Get code examples for specific technologies and use cases
     * 
     * @param {string} technology - Technology name (e.g., 'typescript', 'python', 'react')
     * @param {string} useCase - Specific use case (e.g., 'authentication', 'database-query')
     * @returns {Promise<Object>} Code examples and best practices
     */
    async getExamples(technology, useCase = '') {
        const payload = {
            technology,
            use_case: useCase,
            format: 'code_snippets'
        };
        
        return await this._makeRequest('/examples', payload);
    }

    /**
     * Get best practices for a specific domain or framework
     * 
     * @param {string} domain - Domain area (e.g., 'security', 'performance', 'testing')
     * @param {string} framework - Specific framework (e.g., 'firebase', 'react', 'nodejs')
     * @returns {Promise<Object>} Best practices and guidelines
     */
    async getBestPractices(domain, framework = '') {
        const payload = {
            domain,
            framework,
            include_antipatterns: true
        };
        
        return await this._makeRequest('/best-practices', payload);
    }

    /**
     * Search for solutions to specific error messages
     * 
     * @param {string} errorMessage - The error message or pattern
     * @param {string} technology - Technology context (e.g., 'firebase', 'typescript')
     * @returns {Promise<Object>} Common solutions and troubleshooting steps
     */
    async searchErrors(errorMessage, technology = '') {
        const payload = {
            error: errorMessage,
            technology,
            include_solutions: true,
            include_prevention: true
        };
        
        return await this._makeRequest('/errors', payload);
    }
}

/**
 * Agent integration helper for JavaScript/Node.js agents
 */
class AgentRefToolsHelper {
    constructor(agentName = 'unknown') {
        this.agentName = agentName;
        this.client = new RefToolsClient();
    }

    /**
     * Intelligent lookup based on agent task and required technologies
     * 
     * @param {string} taskDescription - Description of the task
     * @param {Array<string>} technologies - List of technologies (optional)
     * @returns {Promise<Object>} Comprehensive lookup results
     */
    async lookupForTask(taskDescription, technologies = null) {
        if (!technologies) {
            technologies = this._extractTechnologies(taskDescription);
        }

        const results = {
            task: taskDescription,
            agent: this.agentName,
            technologies,
            documentation: {},
            examples: {},
            bestPractices: {}
        };

        for (const tech of technologies) {
            try {
                // Get docs for each technology
                results.documentation[tech] = await this.client.getDocs(tech, taskDescription);
                
                // Get relevant examples
                results.examples[tech] = await this.client.getExamples(tech, taskDescription);
                
                // Get best practices
                results.bestPractices[tech] = await this.client.getBestPractices(tech);
                
            } catch (error) {
                console.warn(`Failed to get info for ${tech}:`, error.message);
                results.documentation[tech] = { error: error.message };
            }
        }

        return results;
    }

    /**
     * Extract likely technologies from task description
     */
    _extractTechnologies(taskDescription) {
        const techKeywords = {
            firebase: ['firebase', 'firestore', 'functions'],
            typescript: ['typescript', 'ts', 'type'],
            javascript: ['javascript', 'js', 'node'],
            react: ['react', 'jsx', 'component'],
            nodejs: ['node', 'express', 'npm'],
            openai: ['openai', 'gpt', 'ai', 'llm']
        };

        const foundTechs = [];
        const taskLower = taskDescription.toLowerCase();

        for (const [tech, keywords] of Object.entries(techKeywords)) {
            if (keywords.some(keyword => taskLower.includes(keyword))) {
                foundTechs.push(tech);
            }
        }

        return foundTechs.length > 0 ? foundTechs : ['general'];
    }
}

// Convenience functions for direct usage
async function queryRef(endpoint, payload = null) {
    const client = new RefToolsClient();
    return await client._makeRequest(endpoint, payload);
}

async function getDocs(library, query = '', maxTokens = 2000) {
    const client = new RefToolsClient();
    return await client.getDocs(library, query, maxTokens);
}

async function searchAPI(apiName, endpointPattern = '', method = '') {
    const client = new RefToolsClient();
    return await client.searchAPI(apiName, endpointPattern, method);
}

async function getExamples(technology, useCase = '') {
    const client = new RefToolsClient();
    return await client.getExamples(technology, useCase);
}

async function getBestPractices(domain, framework = '') {
    const client = new RefToolsClient();
    return await client.getBestPractices(domain, framework);
}

async function searchErrors(errorMessage, technology = '') {
    const client = new RefToolsClient();
    return await client.searchErrors(errorMessage, technology);
}

// Exception classes
class RefToolsConnectionError extends Error {
    constructor(message) {
        super(message);
        this.name = 'RefToolsConnectionError';
    }
}

class RefToolsAPIError extends Error {
    constructor(message) {
        super(message);
        this.name = 'RefToolsAPIError';
    }
}

// Global client instance for easy access
let _defaultClient = null;

function getDefaultClient() {
    if (!_defaultClient) {
        _defaultClient = new RefToolsClient();
    }
    return _defaultClient;
}

// Firebase Cloud Functions integration helpers
const firebaseFunctions = {
    /**
     * Initialize ref-tools client for Firebase Cloud Functions
     * Should be called once at function startup
     */
    initialize() {
        const client = new RefToolsClient();
        global.refClient = client;
        return client;
    },

    /**
     * Get the global ref-tools client instance
     */
    getClient() {
        if (!global.refClient) {
            return this.initialize();
        }
        return global.refClient;
    }
};

// Export for both CommonJS and ES modules
module.exports = {
    RefToolsClient,
    RefToolsConfig,
    AgentRefToolsHelper,
    RefToolsConnectionError,
    RefToolsAPIError,
    queryRef,
    getDocs,
    searchAPI,
    getExamples,
    getBestPractices,
    searchErrors,
    getDefaultClient,
    firebaseFunctions
};

// ES modules export
if (typeof exports !== 'undefined' && exports.__esModule) {
    exports.default = RefToolsClient;
}

// Test the client if run directly
if (require.main === module) {
    (async () => {
        try {
            const client = new RefToolsClient();
            const health = await client.healthCheck();
            console.log('Ref-Tools MCP Health:', health);
            
            // Example usage
            const docs = await getDocs('firebase', 'authentication');
            console.log('Firebase auth docs:', docs.summary || 'No summary available');
            
        } catch (error) {
            console.error('Error testing Ref-Tools client:', error.message);
        }
    })();
}

File: utils/ref_tools_client.py
#!/usr/bin/env python3
"""
Ref-Tools MCP Client for ACIMguide Agent System
Provides token-efficient access to technical documentation for all agents
"""

import os
import json
import logging
import requests
import time
from typing import Dict, List, Optional, Any, Union
from urllib.parse import urljoin
from dataclasses import dataclass

logger = logging.getLogger(__name__)

@dataclass
class RefToolsConfig:
    """Configuration for Ref-Tools MCP client"""
    base_url: str
    port: int = 4102
    transport: str = 'http'
    timeout: int = 20
    max_retries: int = 3
    fallback_urls: List[str] = None

    def __post_init__(self):
        if self.fallback_urls is None:
            self.fallback_urls = [
                "https://ref-tools-mcp.acimguide.com",
                "http://localhost:8080"  # alternative local port
            ]

class RefToolsClient:
    """
    High-level client for Ref-Tools MCP service
    Provides token-efficient documentation access for agents
    """
    
    def __init__(self, config: Optional[RefToolsConfig] = None):
        if config is None:
            port = int(os.getenv('REF_MCP_PORT', 4102))
            host = os.getenv('REF_MCP_HOST', 'localhost')
            base_url = os.getenv('REF_MCP_URL', f'http://{host}:{port}')
            
            config = RefToolsConfig(
                base_url=base_url,
                port=port
            )
        
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'ACIMguide-Agent-System/1.0'
        })

    def _make_request(self, endpoint: str, payload: Optional[Dict] = None, method: str = 'POST') -> Dict[str, Any]:
        """
        Make HTTP request to MCP server with fallback support
        """
        urls_to_try = [self.config.base_url] + self.config.fallback_urls
        
        for attempt, base_url in enumerate(urls_to_try):
            try:
                url = urljoin(base_url, endpoint)
                logger.debug(f"Attempting {method} {url} (attempt {attempt + 1})")
                
                if method.upper() == 'POST':
                    response = self.session.post(
                        url,
                        json=payload,
                        timeout=self.config.timeout
                    )
                else:
                    response = self.session.get(
                        url,
                        timeout=self.config.timeout
                    )
                
                response.raise_for_status()
                return response.json()
                
            except requests.exceptions.RequestException as e:
                logger.warning(f"Request to {base_url} failed: {e}")
                if attempt == len(urls_to_try) - 1:
                    raise RefToolsConnectionError(f"Failed to connect to Ref-Tools MCP service after trying {len(urls_to_try)} endpoints")
                continue
            except Exception as e:
                logger.error(f"Unexpected error with {base_url}: {e}")
                continue

    def health_check(self) -> Dict[str, Any]:
        """Check if the MCP service is healthy and responsive"""
        try:
            return self._make_request('/health', method='GET')
        except Exception as e:
            return {
                'status': 'unhealthy',
                'error': str(e),
                'timestamp': time.time()
            }

    def get_docs(self, library: str, query: str = "", max_tokens: int = 2000) -> Dict[str, Any]:
        """
        Get documentation for a specific library or API
        
        Args:
            library: Name of the library/API (e.g., 'firebase', 'openai', 'react')
            query: Specific search query within the documentation
            max_tokens: Maximum tokens to return (helps with cost optimization)
            
        Returns:
            Dictionary containing relevant documentation snippets
        """
        payload = {
            'library': library,
            'query': query,
            'max_tokens': max_tokens,
            'format': 'json'
        }
        
        return self._make_request('/docs', payload)

    def search_api(self, api_name: str, endpoint_pattern: str = "", method: str = "") -> Dict[str, Any]:
        """
        Search for specific API endpoints and usage patterns
        
        Args:
            api_name: Name of the API (e.g., 'Firebase Admin', 'OpenAI')
            endpoint_pattern: Pattern to search for (e.g., '/users', 'chat/completions')
            method: HTTP method filter (GET, POST, etc.)
            
        Returns:
            API documentation and examples
        """
        payload = {
            'api': api_name,
            'endpoint': endpoint_pattern,
            'method': method.upper() if method else "",
            'include_examples': True
        }
        
        return self._make_request('/api', payload)

    def get_examples(self, technology: str, use_case: str = "") -> Dict[str, Any]:
        """
        Get code examples for specific technologies and use cases
        
        Args:
            technology: Technology name (e.g., 'typescript', 'python', 'react')
            use_case: Specific use case (e.g., 'authentication', 'database-query')
            
        Returns:
            Code examples and best practices
        """
        payload = {
            'technology': technology,
            'use_case': use_case,
            'format': 'code_snippets'
        }
        
        return self._make_request('/examples', payload)

    def get_best_practices(self, domain: str, framework: str = "") -> Dict[str, Any]:
        """
        Get best practices for a specific domain or framework
        
        Args:
            domain: Domain area (e.g., 'security', 'performance', 'testing')
            framework: Specific framework (e.g., 'firebase', 'react', 'nodejs')
            
        Returns:
            Best practices and guidelines
        """
        payload = {
            'domain': domain,
            'framework': framework,
            'include_antipatterns': True
        }
        
        return self._make_request('/best-practices', payload)

    def search_errors(self, error_message: str, technology: str = "") -> Dict[str, Any]:
        """
        Search for solutions to specific error messages
        
        Args:
            error_message: The error message or pattern
            technology: Technology context (e.g., 'firebase', 'typescript')
            
        Returns:
            Common solutions and troubleshooting steps
        """
        payload = {
            'error': error_message,
            'technology': technology,
            'include_solutions': True,
            'include_prevention': True
        }
        
        return self._make_request('/errors', payload)

# Convenience functions for direct usage
def query_ref(endpoint: str, payload: Dict[str, Any] = None) -> Dict[str, Any]:
    """
    Simple function for direct MCP queries
    """
    client = RefToolsClient()
    return client._make_request(endpoint, payload)

def get_docs(library: str, query: str = "", max_tokens: int = 2000) -> Dict[str, Any]:
    """Convenience function for documentation lookup"""
    client = RefToolsClient()
    return client.get_docs(library, query, max_tokens)

def search_api(api_name: str, endpoint_pattern: str = "", method: str = "") -> Dict[str, Any]:
    """Convenience function for API search"""
    client = RefToolsClient()
    return client.search_api(api_name, endpoint_pattern, method)

def get_examples(technology: str, use_case: str = "") -> Dict[str, Any]:
    """Convenience function for code examples"""
    client = RefToolsClient()
    return client.get_examples(technology, use_case)

def get_best_practices(domain: str, framework: str = "") -> Dict[str, Any]:
    """Convenience function for best practices"""
    client = RefToolsClient()
    return client.get_best_practices(domain, framework)

def search_errors(error_message: str, technology: str = "") -> Dict[str, Any]:
    """Convenience function for error resolution"""
    client = RefToolsClient()
    return client.search_errors(error_message, technology)

# Exception classes
class RefToolsConnectionError(Exception):
    """Raised when unable to connect to Ref-Tools MCP service"""
    pass

class RefToolsAPIError(Exception):
    """Raised when MCP service returns an API error"""
    pass

# Global client instance for easy access
_default_client = None

def get_default_client() -> RefToolsClient:
    """Get or create the default client instance"""
    global _default_client
    if _default_client is None:
        _default_client = RefToolsClient()
    return _default_client

# Agent integration helpers
class AgentRefToolsHelper:
    """
    Helper class for integrating Ref-Tools into agent workflows
    """
    
    def __init__(self, agent_name: str = "unknown"):
        self.agent_name = agent_name
        self.client = RefToolsClient()
    
    def lookup_for_task(self, task_description: str, technologies: List[str] = None) -> Dict[str, Any]:
        """
        Intelligent lookup based on agent task and required technologies
        """
        if not technologies:
            technologies = self._extract_technologies(task_description)
        
        results = {
            'task': task_description,
            'agent': self.agent_name,
            'technologies': technologies,
            'documentation': {},
            'examples': {},
            'best_practices': {}
        }
        
        for tech in technologies:
            try:
                # Get docs for each technology
                results['documentation'][tech] = self.client.get_docs(tech, task_description)
                
                # Get relevant examples
                results['examples'][tech] = self.client.get_examples(tech, task_description)
                
                # Get best practices
                results['best_practices'][tech] = self.client.get_best_practices(tech)
                
            except Exception as e:
                logger.warning(f"Failed to get info for {tech}: {e}")
                results['documentation'][tech] = {'error': str(e)}
        
        return results
    
    def _extract_technologies(self, task_description: str) -> List[str]:
        """Extract likely technologies from task description"""
        tech_keywords = {
            'firebase': ['firebase', 'firestore', 'functions'],
            'typescript': ['typescript', 'ts', 'type'],
            'python': ['python', 'py', 'django', 'flask'],
            'react': ['react', 'jsx', 'component'],
            'nodejs': ['node', 'express', 'npm'],
            'openai': ['openai', 'gpt', 'ai', 'llm']
        }
        
        found_techs = []
        task_lower = task_description.lower()
        
        for tech, keywords in tech_keywords.items():
            if any(keyword in task_lower for keyword in keywords):
                found_techs.append(tech)
        
        return found_techs or ['general']

if __name__ == "__main__":
    # Quick test of the client
    try:
        client = RefToolsClient()
        health = client.health_check()
        print(f"Ref-Tools MCP Health: {health}")
        
        # Example usage
        docs = get_docs('firebase', 'authentication')
        print(f"Firebase auth docs: {docs.get('summary', 'No summary available')}")
        
    except Exception as e:
        print(f"Error testing Ref-Tools client: {e}")
